go-controller/pkg/ovn/egressqos.go

```diff
@@ -0,0 +1,909 @@
+			Match:       match,
+			Priority:    r.priority,
+			Action:      map[string]int{nbdb.QoSActionDSCP: r.dscp},
+			ExternalIDs: map[string]string{"EgressQoS": eq.namespace},
+		}
+		qoses = append(qoses, qos)
+	}
+
+	ops, err := libovsdbops.CreateOrUpdateQoSesOps(oc.nbClient, nil, qoses...)
+	if err != nil {
+		return err
+	}
+	allOps = append(allOps, ops...)
+
+	for _, sw := range logicalSwitches {
+		ops, err := libovsdbops.AddQoSesToLogicalSwitchOps(oc.nbClient, nil, sw, qoses...)
+		if err != nil {
+			return err
+		}
+		allOps = append(allOps, ops...)
+	}
+
+	if _, err := libovsdbops.TransactAndCheck(oc.nbClient, allOps); err != nil {
+		return fmt.Errorf("failed to create qos, err: %s", err)
+	}
+
+	eq.stale = false // we can mark it as "ready" now
+	return nil
+}
+
+func generateEgressQoSMatch(eq *egressQoSRule, hashedAddressSetNameIPv4, hashedAddressSetNameIPv6 string) string {
+	var src string
+	var dst string
+
+	switch {
+	case config.IPv4Mode && config.IPv6Mode:
+		src = fmt.Sprintf("(ip4.src == $%s || ip6.src == $%s)", hashedAddressSetNameIPv4, hashedAddressSetNameIPv6)
+	case config.IPv4Mode:
+		src = fmt.Sprintf("ip4.src == $%s", hashedAddressSetNameIPv4)
+	case config.IPv6Mode:
+		src = fmt.Sprintf("ip6.src == $%s", hashedAddressSetNameIPv6)
+	}
+
+	dst = "ip4.dst == 0.0.0.0/0 || ip6.dst == ::/0" // if the dstCIDR field was not set we treat it as "any" destination
+	if eq.destination != "" {
+		dst = fmt.Sprintf("ip4.dst == %s", eq.destination)
+		if utilnet.IsIPv6CIDRString(eq.destination) {
+			dst = fmt.Sprintf("ip6.dst == %s", eq.destination)
+		}
+	}
+
+	return fmt.Sprintf("(%s) && %s", dst, src)
+}
+
+func (oc *Controller) egressQoSSwitches() ([]string, error) {
+	logicalSwitches := []string{}
+
+	// Find all node switches
+	p := func(item *nbdb.LogicalSwitch) bool {
+		// Ignore external and Join switches(both legacy and current)
+		return !(strings.HasPrefix(item.Name, types.JoinSwitchPrefix) || item.Name == "join" || strings.HasPrefix(item.Name, types.ExternalSwitchPrefix))
+	}
+
+	nodeLocalSwitches, err := libovsdbops.FindLogicalSwitchesWithPredicate(oc.nbClient, p)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch local switches for EgressQoS, err: %v", err)
+	}
+
+	for _, nodeLocalSwitch := range nodeLocalSwitches {
+		logicalSwitches = append(logicalSwitches, nodeLocalSwitch.Name)
+	}
+
+	return logicalSwitches, nil
+}
+
+type mapOp int
+
+const (
+	mapInsert mapOp = iota
+	mapDelete
+)
+
+type mapAndOp struct {
+	m  *sync.Map
+	op mapOp
+}
+
+func (oc *Controller) syncEgressQoSPod(key string) error {
+	startTime := time.Now()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	klog.Infof("Processing sync for EgressQoS pod %s/%s", namespace, name)
+
+	defer func() {
+		klog.V(4).Infof("Finished syncing EgressQoS pod %s on namespace %s : %v", name, namespace, time.Since(startTime))
+	}()
+
+	obj, loaded := oc.egressQoSCache.Load(namespace)
```

Comment:
yeah when we process a qos we add all of the relevant pods at that point to the address set, so it would be added there if we skipped it in the pod path.
about a pod not having an ip that sould be covered in the update, should I add a unit test for that?